libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,892 @@
+        break;
+    case Event::EXIT_MENU:
+        close_5_key_OSD_cable_connection(FUNCTOR_BIND_MEMBER(&AP_RunCam::handle_5_key_simulation_response, void, const Response&));
+        break;
+    case Event::IN_MENU_UP:
+    case Event::IN_MENU_RIGHT:
+    case Event::IN_MENU_DOWN:
+    case Event::IN_MENU_ENTER:
+    case Event::IN_MENU_EXIT:
+        simulate_5_key_OSD_cable_button_press(map_key_to_protocol_operation(key), FUNCTOR_BIND_MEMBER(&AP_RunCam::handle_5_key_simulation_response, void, const Response&));
+        break;
+    case Event::BUTTON_RELEASE:
+        simulate_5_key_OSD_cable_button_release(FUNCTOR_BIND_MEMBER(&AP_RunCam::handle_5_key_simulation_response, void, const Response&));
+        break;
+    case Event::NONE:
+    default:
+        break;
+    }
+}
+
+// every time we run the OSD menu simulation it's necessary to open the connection
+void AP_RunCam::open_5_key_OSD_cable_connection(parse_func_t parseFunc)
+{
+    send_request_and_waiting_response(Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION,
+        uint8_t(ConnectionOperation::RCDEVICE_PROTOCOL_5KEY_FUNCTION_OPEN), 400, 2, parseFunc);
+}
+
+// every time we exit the OSD menu simulation it's necessary to close the connection
+void AP_RunCam::close_5_key_OSD_cable_connection(parse_func_t parseFunc)
+{
+    send_request_and_waiting_response(Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION,
+        uint8_t(ConnectionOperation::RCDEVICE_PROTOCOL_5KEY_FUNCTION_CLOSE), 400, 2, parseFunc);
+}
+
+// simulate button press event of 5 key OSD cable with special button
+void AP_RunCam::simulate_5_key_OSD_cable_button_press(const SimulationOperation operation, parse_func_t parseFunc)
+{
+    if (operation == SimulationOperation::SIMULATION_NONE) {
+        return;
+    }
+
+    send_request_and_waiting_response(Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_PRESS, uint8_t(operation), 400, 2, parseFunc);
+}
+
+// simulate button release event of 5 key OSD cable
+void AP_RunCam::simulate_5_key_OSD_cable_button_release(parse_func_t parseFunc)
+{
+    send_request_and_waiting_response(Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_RELEASE,
+        uint8_t(SimulationOperation::SIMULATION_NONE), 400, 2, parseFunc);
+}
+
+// send a RunCam request and register a response to be processed
+void AP_RunCam::send_request_and_waiting_response(Command commandID, uint8_t param,
+    uint32_t timeout, uint16_t maxRetryTimes, parse_func_t parserFunc)
+{
+    drain();
+
+    Response response(this, commandID, param, timeout, maxRetryTimes, parserFunc);
+
+    if (_waiting_response_queue.push(response)) {
+        debug("sending command: %d, op: %d\n", int(commandID), int(param));
+        // send packet
+        send_packet(commandID, param);
+    }
+}
+
+// send a packet to the serial port
+void AP_RunCam::send_packet(Command command, uint8_t param)
+{
+    // is this device open?
+    if (!uart) {
+        return;
+    }
+
+    uint8_t buffer[4];
```

Comment:
I'd prefer to see this buffer assembled from its components statically rather than using the `*b++` stuff.

Reply:
Fixed